app/ante/gov.go

```diff
@@ -2,44 +2,51 @@ package ante
 
 import (
 	"cosmossdk.io/errors"
-	gogoany "github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	"github.com/cosmos/cosmos-sdk/x/authz"
-	gov "github.com/cosmos/cosmos-sdk/x/gov/types"
 	govv1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
-	"github.com/cosmos/cosmos-sdk/x/params/types/proposal"
-	gogoproto "github.com/cosmos/gogoproto/proto"
 )
 
+type ParamFilter func(sdk.Msg) error
+
 // GovProposalDecorator ensures that a tx with a MsgSubmitProposal has at least one message in the proposal.
 // Additionally it replace the x/paramfilter module that existed in v3 and earlier versions.
 type GovProposalDecorator struct {
-	// forbiddenGovUpdateParams is a map of type_url to a list of parameter fields that cannot be changed via governance.
-	forbiddenGovUpdateParams map[string][]string
+	// paramFilters is a map of type_url to a list of parameter fields that cannot be changed via governance.
+	paramFilters map[string]ParamFilter
 }
 
-func NewGovProposalDecorator(forbiddenParams map[string][]string) GovProposalDecorator {
-	return GovProposalDecorator{forbiddenParams}
+func NewGovProposalDecorator(paramFilters map[string]ParamFilter) GovProposalDecorator {
+	return GovProposalDecorator{
+		paramFilters: paramFilters,
+	}
 }
 
 // AnteHandle implements the AnteHandler interface.
 // It ensures that MsgSubmitProposal has at least one message
 // It ensures params are filtered within messages
 func (d GovProposalDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
 	for _, m := range tx.GetMsgs() {
-		if proposal, ok := m.(*govv1.MsgSubmitProposal); ok {
-			if len(proposal.Messages) == 0 {
```

Comment:
This was part of v3, I think we should keep it: https://github.com/01builders/celestia-app/blob/dff432201cebaa0f3424e6018f7de2d3cc1d86b9/app/ante/gov.go#L23

Reply:
it still errors out in `checkNestedMsgs` just below on len(0)